### DISCUSSION TITLE
Implement MongoDB & Redis API?

### DISCUSSION THREAD

**Discussion by asapostolov** [ravendb/ravendb#14049] [Ideas]
I was thinking how to promote RavenDB to projects that are using MongoDB and was thinking how cool it would be to be able to switch an entire project from Mongo to Raven and compare parameters like direct performance, ease of development for new features etc. with minimal overhead for rewriting the data layer.

I saw that you guys have implemented import of data from MongoDB in the import section of Raven (I haven't checked how that works in practice, but presume it does). Was curious how hard it is to implement the entire Mongo API, so potentially a project could move its config from Mongo to Raven server and it will Just Work. 

I presume this would have a huge cost in terms of complexity and maintenance, but it could also open up RavenDB for much wider adoption if it can outperform as a direct plug-and-play competitor to Mongo.

Also - Redis - it's mostly used for caching and it's widely available for most major hosting platforms, but I have 2 projects where the high performance cache aspect of Redis is not that important (I presume Redis would have much better performance from Raven for what it does). In these cases it's mostly just used for enabling libraries like hangfire, ASP.NET Cache (where performance is somewhat important) or elmah. Might be cool to be able to use RavenDB as a plug-and-play replacement in these cases.
You can also try to optimize Raven to have Redis-level performance in certain special cases which would be a curious use-case/alternative.

Have you guys on the RavenDB team thought about any of these? I know all this might sound a bit hilarious but it might also be a cool thought experiment to explore or prototype.

**Comment by ayende** [comment]
Implementing the MongoDB API directly would be a _huge_ task. Without going too deep, assume a full team for a minimum of 6 months, more likely a year+ to do so.
There are licensing issues with that, too, which likely put us on MongoDB 3.6 compatibility by default (haven't looked long enough to give a good answer).

A key issue here is that the MongoDB API requires to do things in a certain way, and it is pretty complex in how it does so.

Implementing a Redis protocol, on the other hand, is far simpler.
The API is straightforward and quite easy to translate in to RavenDB terms.

However, there is another issue here, which is the security model (as is, there is none):
https://redis.io/docs/manual/security/#security-model

There seems to be TLS / mutual auth since the last time I checked, but they aren't well supported.
For example, both Hangfire and StackExchange.Redis do not appear to support authentication Redis.

**Comment by ayende** [comment]
I do think that a Redis layer would be a cool hack, and on its face, it doesn't seem like a hard thing to do. 
If you are interested, we will be very happy to assist in this
